For final review of Gardener installation documentation

je-hal opened the pull request:
Although we had a successful test run we'd like to have a final review of this documentation before merging it here.

See [preview](https://github.com/JensHaley/garden-setup) in my fork.

@Diaphteiros : Feel free to add comments before our meeting on Monday.

je-hal committed:
moving new documentation for gardener installation

je-hal committed:
adding latest changes

je-hal committed:
Updating README.md

je-hal committed:
Removed a step "export KUBECONFIG="

This step was already part of the prerequisites

je-hal committed:
deleted space

je-hal committed:
Correction

je-hal committed:
Update README.md

Diaphteiros commented on README.md:
````diff
@@ -1,3 +1,273 @@
+    watch -d kubectl -n garden get pods,ingress,sts,svc
+    ```
+
+1. In your first terminal window, use the following command to deploy Gardener on your base cluster:
````
It's probably a good idea to mention a `sow order -A` before. The command will detect some kinds of errors in the acre.yaml (if the user made some), but it doesn't deploy something. Also, the user than has an idea about the amount and order of components.

JensHaley replied:
done

Diaphteiros commented on README.md:
````diff
@@ -1,3 +1,273 @@
+```yaml
+etcd:
+  backup:
+    type: <gcs|s3>                    # type of blob storage
````
For Azure: add `etcd.backup.resourceGroup`

JensHaley replied:
Added in the documentation for `landscape.etcd`, but I'd need more information what this parameter is used for and where it is found. Is that value coming from Azure or is it a fixed value defined somewhere in `sow`?

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+| Field | Type | Description | Example |Iaas Provider Documentation |
+|:------|:--------|:--------|:--------|:--------|
+|`region`|IaaS provider specific|Region where Gardener will create seed clusters and shoot clusters. | `europe-west1`|[GCP (HowTo)](https://cloud.google.com/kubernetes-engine/docs/how-to/managing-clusters#viewing_your_clusters), [GCP (overview)](https://cloud.google.com/docs/geography-and-regions); [AWS (HowTo)](https://docs.aws.amazon.com/cli/latest/reference/eks/describe-cluster.html), [AWS (Ov ...
+|`zones`|IaaS provider specific|Zones where Gardener will create seed clusters and shoot clusters. |`europe-west1-b`|[GCP (HowTo)](https://cloud.google.com/kubernetes-engine/docs/how-to/managing-clusters#viewing_your_clusters), [GCP (overview)](https://cloud.google.com/docs/geography-and-regions); [AWS (HowTo)](https://docs.aws.amazon.com/cli/latest/reference/eks/describe-cluster.html), [AWS (Ove ...
```
For Azure: remove `zones` completely.

JensHaley replied:
note: remove complete zones: block for Azure

JensHaley replied:
done

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+| Field | Type | Description | Example&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Iaas Provider Documentation |
+|:------|:--------|:--------|:--------|:---------|
+|`backup.type`|Fixed value| Type of your blob store. Supported blob stores: `gcs` ([Google Cloud Storage](https://cloud.google.com/storage/)), and `s3` ([Amazon S3](https://aws.amazon.com/s3/)).|`gcs`|n.a.|
```
supported Azure backup: `abs`

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+| Field | Type | Description | Example |IaaS Provider Documentation
+|:------|:--------|:--------|:--------|:------------|
+|`type`|Fixed value|Your DNS provider. Supported providers: `google-clouddns` ([Google Cloud DNS](https://cloud.google.com/dns/docs/)), and `aws_route53` ([Amazon Route 53](https://aws.amazon.com/route53/)).|`google-clouddns`|n.a.|
```
It's `aws-route53`, not `aws_route53`. `azure-dns` is also supported from garden-setup, but not yet from gardener itself (and thus can't be used here). This should change when https://github.com/gardener/gardener/pull/850 is merged.

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+| Field | Type | Description | Example |IaaS Provider Documentation
+|:------|:--------|:--------|:--------|:------------|
+|`type`|Fixed value|Your DNS provider. Supported providers: `google-clouddns` ([Google Cloud DNS](https://cloud.google.com/dns/docs/)), and `aws_route53` ([Amazon Route 53](https://aws.amazon.com/route53/)).|`google-clouddns`|n.a.|
+|`hostendZoneID`| k8s domain| The domain of your base cluster (replace dots with dashes).|`gcp.dev.k8s.jacksgrocerystore.com`|[GCP](https://cloud.google.com/sdk/gcloud/reference/dns/managed-zones/list), [AWS](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/ListInfoOnHostedZone.html)|
```
Probably not needed anymore after https://github.com/gardener/gardener/pull/850 is merged.

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+| `state` | Important state information of the components is stored here, e.g. the terraform state and generated certificates. It is crucial that this directory is not deleted while the landscape is active. While the contents of the *export* and *gen* folders will be overwritten when a component is deployed again, the contents of *state* will be reused instead. In some cases, it is necessary to d ...
+
+#### Deleting Directories
+The `gen` directory can be safely deleted at any time, except *during* the deployment of a component.
```
This is no longer true.

`gen` and `export` can be re-created from scratch by re-deploying the components, `gen` cannot. 

However, deleting a component might files that are created on deployment and stored in `gen`, so when the `gen` folder is deleted, it might be necessary to re-create it before it is possible to delete the component.

Also, all three folders (or their contents, at least) are deleted upon component deletion. Manually deleting them is only necessary in special cases.

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+        hash:                                 # bcrypted hash of password, see above
+</pre>
+
+
```
We are currently missing the formats (they yaml keys, to be more precise) in which the credentials need to be specified. These are:

AWS:
region: ...
accessKeyID: ...
secretAccessKey: ...

GCP:
serviceaccount.json: |
  {
    "type": "...",
    "project_id": "...",
    ...
  }

Azure:
clientID: ...
clientSecret: ...
subscriptionID: ...
tenantID: ...

JensHaley replied:
done

Diaphteiros reviewed (commented):
I added some comments regarding missing or outdated information.

afritzler commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+<pre>
+landscape:
+  <a href="#landscapename">name</a>: &lt;URL-friendly name&gt;                       # Gardener managed k8s landscape name
```
I would use here something else as an example. Instead of "URL friendly" we should use something like "my-gardener".

JensHaley replied:
I am not convinced ;-)
My intention was to give readers a hint where this name will be used later.

What is more, `my-gardener` is an example, not a description of the field. If we use an example, we should omit the brackets (&lt; &gt;), but then it looks like a constant value.

Proposal:
```yaml
landscape:
  name: <Identifier>                  # general landscape identifier, for example, `my-gardener`
```

urabenst commented on README.md:
```diff
@@ -1,3 +1,273 @@
+  backup:
+    type: <gcs|s3>                    # type of blob storage
+    region: (( iaas.region ))         # region of blob storage (default: same as above)
+    credentials: (( iaas.creds ))     # credentials for the blob storage's IaaS provider (default: same as above)
```
Should this be "(( iaas.credentials ))"?

JensHaley replied:
Yes, thanks!

JensHaley replied:
done

urabenst commented on README.md:
```diff
@@ -1,3 +1,273 @@
+* You need to have the following tools installed:
+  * [Docker](https://docs.docker.com/install/)
+  * [git](https://git-scm.com/downloads)
+  * [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
```
One also needs `make` and `yaml2json` if you want to include that.

Diaphteiros replied:
The sow image is now available in a registry, so `make` is only needed if you want to build it yourself. 
We should be able to get rid of the `yaml2json`, I think.

Diaphteiros replied:
With https://github.com/gardener/garden-setup/commit/d31f816f31e848331baab5daace5d92fef5d7203 the `yaml2json` dependency (and one to `jq`) should be gone.

JensHaley replied:
Isn't `make` still required to build to Docker image?

urabenst commented on README.md:
```diff
@@ -1,3 +1,273 @@
+    # setup for calling sow via the wrapper
+    git clone "https://github.com/gardener/sow"
+    cd sow
+    make # will build and tag the image
```
The user needs to be in the Docker group or one has to run `sudo make` here.

Diaphteiros replied:
We now push the image to a registry, so the `make` step has become optional. Should still be mentioned in case someone needs to build the image manually for whatever reason.

JensHaley replied:
Then we should mention the registry. Which one is it?

JensHaley replied:
I rewrote this part referring to  eu.gcr.io/gardener-project/sow

urabenst commented on README.md:
````diff
@@ -1,3 +1,273 @@
+```
+CIDR ranges of your base cluster.
+
+Finding out CIDR ranges of your cluster is not trivial. For example, GKE only tells you a "pod address range" which is actually a combination of pod and service CIDR. However, since the `kubernetes` service typically has the first IP of the service IP range and most methods to get a kubernetes cluster tell you at least something about the CIDRs, it is usually possible to find out the CIDRs with a ...
````
It took us a while to find the correct CIDRs for Azure. So maybe, we could extend the documentation here?
For AKS, one can retrieve the node CIDR using 
`az network vnet list -g MC_${CLUSTER_NAME}_${CLUSTER_NAME}_$REGION -o json | jq -r '.[] | .subnets[] | .addressPrefix'` 
and the pod CIDR with
`kubectl -n kube-system get daemonset/kube-proxy -o yaml | grep cluster-cidr= | grep -v annotations | cut -d = -f2'`.
As a service CIDR we used `10.0.0.0/16`
Do you plan to include similar commands for AWS or GCP?

JensHaley replied:
Thank you for your feedback. I agree that this is the tricky part.

For the first roll-out, it is not planned yet to provide these details. It may be useful to keep provider specific parts of the configuration in a separate document (tutorials?).

urabenst reviewed (commented):
Here are some suggestions from our side. Hope it helps

Diaphteiros commented on README.md:
````diff
@@ -1,3 +1,273 @@
+generating exports
+*** species dashboard deployed
+```  
+
````
Maybe we should change this. The dashboard is not necessarily the last component that is deployed, so this might be confusing. 

Furthermore, it is now possible to also get the dashboard url by calling `sow url` (or reading the `dashboard_url` file in the dashboard's export folder).

JensHaley replied:
done

Diaphteiros commented on README.md:
````diff
@@ -1,3 +1,273 @@
+        gcloud container clusters get-credentials <your_cluster> --zone <your_zone> --project <your_project>
+        ```
+
+    1. Call `crop/bin/convertkubeconfig 3.1`. When asked for credentials, enter the ones that the GKE dashboard shows when clicking on `show credentials`.
````
This has changed now. The GKE kubeconfig is expected at the path specified in the acre.yaml (`landscape.cluster.kubeconfig`) and will be **replaced** with the basic-auth variant using the command `sow convertkubeconfig`. The bin/convertkubeconfig script has been removed to reduce the dependencies.

JensHaley replied:
done

Diaphteiros commented on README.md:
```diff
@@ -1,3 +1,273 @@
+  <a href="#landscapename">name</a>: &lt;URL-friendly name&gt;                       # Gardener managed k8s landscape name
+
+  <a href="#landscapecluster">cluster</a>:                                          # Information about your base cluster
+    kubeconfig: &lt;relative path + filename&gt;          # Path to your `kubeconfig` file, rel. to folder `landscape`
```
Since the docker container that runs `sow` only mounts parts of the surrounding file system, absolute paths are actually not supported here. To avoid problems, the kubeconfig should be within the landscape directory (or a subdirectory of it).

JensHaley replied:
Not sure what you mean here. It says `relative path` above, so all should be fine, right?

je-hal committed:
sow order -A + backup.resourceGroup

JensHaley commented on README.md:
```diff
@@ -1,3 +1,273 @@
+    services: &lt;CIDR IP range&gt;                             
+
+  <a href="#landscapeiaas">iaas</a>:
+    region: &lt;major region&gt;-&lt;minor region&gt;      # region that Gardener will use for seed clusters
```
Azure seems to have region names with spaces like "West Europe", "East US", see [here](https://azure.microsoft.com/en-us/global-infrastructure/geographies/).

Is `westeurope` a technical name of Azure that is available in the Azure service account or do you do some kind of conversion to deal with the spaces?

urabenst replied:
Both works with the CLI, but for technical purposes they use the names in small caps without spaces. You can see that for instance with `az account list-locations`:

![image](https://user-images.githubusercontent.com/39942780/55719792-63f45400-59ff-11e9-81ab-d782afe78e7e.png)

JensHaley replied:
@urabenst : Thank you for your feedback. For now I'd recommend for Azure to convert the names given  in the [region list](https://azure.microsoft.com/en-us/global-infrastructure/geographies/): remove spaces, and convert capital letters to lower letters. If you know any command or list I could refer to instead, that would be great.

urabenst replied:
I would suggest to use the command above to retrieve the regions names. Anyone who wants to deploy Gardener on Azure, will have access to the CLI.

je-hal committed:
test

je-hal committed:
Nearly done

Open: directories, more links to Azure documentation, Azure regions

je-hal committed:
Minor corrections - sorted sow commands

je-hal committed:
Azure links and new section to uninstall

Diaphteiros reviewed (changes requested):
Apart from the two small things I mentioned in the comments, this looks great!

JensHaley commented on README.md:
````diff
@@ -1,3 +1,295 @@
+        ```
+
+    1. Save the GKE kubeconfig in a local file (for example `kubekonfig`) in your `landscape` folder. Make sure that you assign this file to field `landscape.cluster.kubeconfig` in the configuration file `acre.yaml` that is created in the next step.
+    > The filename and path you will use in file `acre.yaml` should be *different* from the file you have set in your `KUBECONFIG` environment variable.
````
@Diaphteiros: Wouldn't it make more sense for the user then to set the KUBECONFIG variable to the actual `kubeconfig` file that is overwritten to access the cluster? Or does the variable KUBECONFIG still have to point to a file that does not exist before you execute `sow covertkubeconfig`?

If KUBECONFIG may reference the created file I would get rid of sub-step 1, and instead of the comment I'd add a step that asks users to set the KUBECONFIG variable to the created file.

Thanks

Diaphteiros replied:
The `KUBECONFIG` variable isn't evaluated at all in `sow convertkubeconfig`. It doesn't matter where it points to. 
Actually, since the `sow` command is run inside a docker container, it has a completely different `KUBECONFIG` anyway.
So this can be simplified to pointing the `KUBECONFIG` to the (non-existing) `landscape/kubeconfig`, then the `gcloud` command and then `sow convertkubeconfig` (GKE scenario).

JensHaley replied:
@Diaphteiros : Ok, understood.

Nevertheless, all users need to set their `KUBECONFIG` environment variable to be able to do the `watch kubectl` later. 

They might have done that in some shell-rc file, but to be sure no one stumbles over this I added it as an explicit step (step 7), and added a recommendation to save it in your landscape directory (step 4). 

That way, we solve the "hen and egg" problem and the only GKE specific step that remains really is step 6. 

Hope you like my reorganization - otherwise let's have a short call.

mandelsoft commented on README.md:
```diff
@@ -1,3 +1,295 @@
+  <a href="#landscapeetcd">etcd</a>:
+    backup:
+      type: &lt;gcs|s3&gt;                          # type of blob storage
+      resourceGroup:                          # Azure specific, see below
```
if you have a see below here, you should also hyperlink the `resourceGroup` key

JensHaley replied:
I thought it would be sufficient to have the key to the `etcd` key.

Anyway, I will change the comment to:
`# Azure resource group you would like to use for your backup`

Then it's consistent with the description in the `etcd` block I link to and users do not have to follow a link just for key `resourceGroup` - it would be difficult to link to that entry in my table.

mandelsoft commented on README.md:
```diff
@@ -1,3 +1,295 @@
+| Field | Type | Description | Example |
+|:------|:--------|:--------|:--------|
+|`kubeconfig`|File path| Path to your kubeconfig, relative to your landscape folder. It is recommended to create a kubeconfig file in your landscape folder to be able to sync all files relevant for your installation with a git repository.| `./kubeconfig` |
+|`domain`| Unique name| Basis domain for DNS entries. As a best practice, use an individual prefix together with the cluster domain of your base cluster.|`vedge.gcp.dev.k8s.jacksgrocerystore.com`|
```
It is important to note here that access to the appropriate hosted zone must be available. Because with the base domain other dns entries will be created. May be a reference zo the dns section would be useful.

JensHaley replied:
@mandelsoft @Diaphteiros :
Diaphteiros says below that after the merge of [#850](https://github.com/gardener/gardener/pull/850) the key `hostedZoneID` might no longer be needed.

Will this comment still be relevant if we remove `hostedZoneID` ? - Thanks.

je-hal committed:
Reorganized the KUBECONFIG information

The conversion of file `kubeconfig` is independent of what KUBECONFIG points to. Nevertheless - and even if users do not use GKE - it needs to be set at some point, so to simplify it this step was added and for GKE an additional step was added.

mandelsoft commented on README.md:
````diff
@@ -1,3 +1,295 @@
+    - <major region>-<minor region>-<zone>          # Example: europe-west1-d
+  credentials:                                      # credentials to get access to the seed cluster through service account
+```
+Contains the information where Gardener will create seed clusters. By default, the *initial* seed component will create a seed resource using your base cluster as seed cluster.
````
The use of *seed* clusters here (and some lines above and below) is wrong. It is used to configure an initial seed and default  profile used to create *shoot* clusters. 

Other seeds and profiles can be added later on.

JensHaley replied:
@mandelsoft: 
I hopefully found all cases and explain it now in the way you described it above.

Could you give me an idea what "profiles" are? Is that something users will be able to see and edit on the Gardener dashboard after the installation?

Diaphteiros replied:
"profile" means a `Cloudprofile" (e.g. https://github.com/gardener/gardener/blob/master/example/30-cloudprofile-gcp.yaml). Cloudprofiles are not directly visible in the Gardener Dashboard, but they define the constraints for shoots. For example, you can only choose a specific kubernetes version for your shoot, if that version is contained in a Cloudprofile. To link it to the Dashboard, the Cloudprofiles define which values can be chosen in the drop-down menus.

mandelsoft commented on README.md:
````diff
@@ -1,3 +1,295 @@
+### landscape.iaas
+```yaml
+iaas:
+  region: <major region>-<minor region>             # region that Gardener will use for seed clusters
````
see remark below

je-hal committed:
added missing brackets

mandelsoft commented on README.md:
```diff
@@ -1,3 +1,295 @@
+|`zones`|IaaS provider specific|Zones where Gardener will create seed clusters and shoot clusters. This block is only required for GCP or AWS. |`europe-west1-b` (GCP)<br/></br>|[GCP (HowTo)](https://cloud.google.com/kubernetes-engine/docs/how-to/managing-clusters#viewing_your_clusters), [GCP (overview)](https://cloud.google.com/docs/geography-and-regions); [AWS (HowTo)](https://docs.aws.amazon.co ...
+|`credentials`|IaaS provider specific|Service account credentials in a provider-specific format. | See table with yaml keys below. | [GCP](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating_service_account_keys), [AWS](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html#id_users_service_accounts), [Azure](https://docs.microsoft.com/en-us/azure/aks/kuberne ...
+
+The service account credentials will be used to give Gardener access to your base cluster:
```
The *credentials* (service account is only for GKE) provide access to the iaas layer used for creating required resources for *shoot* clusters. They are *not* used to access the base cluster.

JensHaley replied:
@mandelsoft : 
I rewrote it, but I have a feeling that further clarification is needed.

What exactly is the "IaaS layer"? It seems to be something different for each IaaS provider so it might be helpful to name for each provider what is meant.

- GKE: service account
- AWS: ???
- Azure: ???

I am not even sure any more I linked to the right documentation here. If I new what 'IaaS layer' means in each case I could easier find a useful link in the IaaS provider's documentation.

je-hal committed:
minor correction

JensHaley commented on README.md:
```diff
@@ -1,3 +1,295 @@
+
+| Field | Type | Description | Example&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Iaas Provider Documentation |
+|:------|:--------|:--------|:--------|:---------|
+|`backup.type`|Fixed value| Type of your blob store. Supported blob stores: `gcs` ([Google Cloud Storage](https://cloud.google.com/storage/)), `s3` ([Amazon S3](https://aws.amazon.com/s3/)), and `abs`[Azure Blob Storage](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-overview).|`gcs`|n.a.|
```
done

mandelsoft commented on README.md:
````diff
@@ -1,3 +1,295 @@
+  hostedZoneID:                             # hosted zone id for chosen domain
+  credentials:                              # credentials for the dns provider
+```
+Configuration for the Domain Name Service (DNS) provider. If your IaaS provider offers a DNS you can use the same values for `dns.credentials` as for `iaas.creds` above by using the [(( foo ))](https://github.com/mandelsoft/spiff/blob/master/README.md#-foo-) expression of spiff.
````
If the hosted zone for your selected dns base name (in cluster) is owned by your IAAS account, the same credentials can be used. If the belong to another account (or another iass provider) the appropriate credentials (and type) have to be configured.

JensHaley replied:
ok, thanks, I made this more explicit now.

JensHaley commented:
Brackets were added, information around `kubeconfig` conversion reorganized, see above.

je-hal committed:
2 additions for mandelsoft

Diaphteiros commented on README.md:
````diff
@@ -1,3 +1,296 @@
+    gcloud container clusters get-credentials <your_cluster> --zone <your_zone> --project <your_project>
+    ```
+
+1. It is recommended to save your `kubeconfig` in your `landscape` directory. For the remaining steps we will assume that you saved it using file path `landscape/kubeconfig`.
````
I'd probably rephrase this. This way, it sounds like it would be a good idea to put the kubeconfig in the landscape directory, but it's not necessary. Due to the filesystem mounting in the docker container, the kubeconfig actually *has to be* somewhere inside the landscape folder (or within a parent directory of it). For simplicity, we should just write that the kubeconfig needs to be somewhere within the landscape directory, so no-one tries to put it elsewhere and wonders why the setup doesn't work.

JensHaley replied:
ok, it now says: "Save your `kubeconfig` somewhere in your `landscape` directory, leaving out the "It is recommended.

je-hal committed:
seeds and credentials - refinements

je-hal committed:
corrected typo

je-hal committed:
save location of kubeconfig
